### PR TITLE
feat(apigatewayv2): WebSocket API - IAM authorizer support

### PR DESCRIPTION
**Pull request by d0z0**
This adds support for `AWS_IAM` as Authorizer for Websocket $connect route.

The CDK supports adding IAM Authorizer as `authorizationType` for `HttpApi`, but does not support it for `WebSocketApi` L2 construct

IAM Authorization is covered in the docs [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-control-access-iam.html). 

It works the same way as REST or HTTP API's where you can make an endpoint (connect route for websocket) publicly inaccessible, and setup an IAM user, and allow access using signed URL's

The above doc links back to [this](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html) section which explains the `AWS_IAM`

This is also available from the AWS console for Websocket connect Route:

Screenshot 2022-07-30 at 17 47 51

---

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigatewayv2-authorizers/README.md [hunk 1]
```diff
@@ -256,3 +257,24 @@ new apigwv2.WebSocketApi(this, 'WebSocketApi', {
   },
 });
 ```
+
+### IAM Authorizer
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apigatewayv2-authorizers/README.md:261]: There is already a section in this README with this name.

File: packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts [hunk 2]
```diff
@@ -161,14 +169,18 @@ export interface IWebSocketRouteAuthorizer {
...
 }
 
 /**
  * Explicitly configure no authorizers on specific WebSocket API routes.
  */
 export class WebSocketNoneAuthorizer implements IWebSocketRouteAuthorizer {
-  public bind(_: WebSocketRouteAuthorizerBindOptions): WebSocketRouteAuthorizerConfig {
+  public bind(
+    _: WebSocketRouteAuthorizerBindOptions,
+  ): WebSocketRouteAuthorizerConfig {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts:183]: please revert this style change.

File: packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts [hunk 3]
```diff
@@ -161,14 +169,18 @@ export interface IWebSocketRouteAuthorizer {
   /**
    * Bind this authorizer to a specified WebSocket route.
    */
-  bind(options: WebSocketRouteAuthorizerBindOptions): WebSocketRouteAuthorizerConfig;
+  bind(
+    options: WebSocketRouteAuthorizerBindOptions
+  ): WebSocketRouteAuthorizerConfig;
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts:174]: please revert this style change.

File: packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts [hunk 4]
```diff
@@ -72,18 +74,24 @@ export interface WebSocketAuthorizerAttributes {
...
+  implements IWebSocketAuthorizer {
   /**
    * Import an existing WebSocket Authorizer into this CDK app.
    */
-  public static fromWebSocketAuthorizerAttributes(scope: Construct, id: string, attrs: WebSocketAuthorizerAttributes): IWebSocketRouteAuthorizer {
+  public static fromWebSocketAuthorizerAttributes(
+    scope: Construct,
+    id: string,
+    attrs: WebSocketAuthorizerAttributes,
+  ): IWebSocketRouteAuthorizer {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts:94]: please revert this style change.

File: packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts [hunk 5]
```diff
@@ -72,18 +74,24 @@ export interface WebSocketAuthorizerAttributes {
...
 }
 
 /**
  * An authorizer for WebSocket Apis
  * @resource AWS::ApiGatewayV2::Authorizer
  */
-export class WebSocketAuthorizer extends Resource implements IWebSocketAuthorizer {
+export class WebSocketAuthorizer
+  extends Resource
+  implements IWebSocketAuthorizer {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts:86]: please revert this style change.

File: packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts [hunk 6]
```diff
@@ -12,6 +12,9 @@ import { IWebSocketRoute } from './route';
 export enum WebSocketAuthorizerType {
   /** Lambda Authorizer */
   LAMBDA = 'REQUEST',
+
+  /** IAM Authorizer */
+  IAM = 'AWS_IAM',
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts:17]: what does this do? Can you provide a link to the docs that describe what this does?

**Review by comcalvi — CHANGES_REQUESTED** [review]
I don't understand what functionality this adds. From our existing docs (the readme), it looks like this is already supported.

File: packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts (hunk 6 above)
Comment (d0z0) [reply packages/@aws-cdk/aws-apigatewayv2/lib/websocket/authorizer.ts:17]: Thanks for reviewing. So this is covered in the docs [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-control-access-iam.html). 

It works the same way as REST or HTTP API's where you can make an endpoint (connect route for websocket) publicly inaccessible, and setup an IAM user, and allow access using signed URL's

The above doc links back to [this](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html) section which explains the `AWS_IAM`

File: packages/@aws-cdk/aws-apigatewayv2-authorizers/README.md (hunk 1 above)
Comment (d0z0) [reply packages/@aws-cdk/aws-apigatewayv2-authorizers/README.md:261]: Sorry, should have made it more obvious. The other one is under `HTTP` API's and this one is under `WebSocket` and uses the singular word "Authorizer". I can make them separate but thought I'll copy the style that's been followed 😄

**Comment by d0z0** [comment]
> I don't understand what functionality this adds. From our existing docs (the readme), it looks like this is already supported.

Thanks for the review. I should have provided more context, I've now updated the description as well

File: packages/@aws-cdk/aws-apigatewayv2-authorizers/README.md [hunk 7]
```diff
@@ -256,3 +257,24 @@ new apigwv2.WebSocketApi(this, 'WebSocketApi', {
...
+// This function handles your connect route
+declare const connectHandler: lambda.Function;
+
+const webSocketApi = new apigwv2.WebSocketApi(this, 'WebSocketApi');
+
+webSocketApi.addRoute('$connect', {
+  integration: new WebSocketLambdaIntegration('Integration', connectHandler),
+  authorizer: new WebSocketIamAuthorizer()
+});
+
```
Comment (corymhall) [inline packages/@aws-cdk/aws-apigatewayv2-authorizers/README.md:278]: Can you add an example here of granting access to the route?

**Review by corymhall — CHANGES_REQUESTED** [review]
Just a minor comment on the readme and then I'm good with this

**Review by corymhall — APPROVED** [review]

**Pull request merged**